File: amongos/amongos.py
'''
Trabalho Em Trenós - PAA ENPE 2020

Thiago de Moraes Teixeira
RA: 760667
'''

import sys 
import numpy as np
import time
  
class Graph(): 
  
    def __init__(self, vertices): 
        self.V = vertices 
        self.graph = [[0 for column in range(vertices)]  
                    for row in range(vertices)] 
  
    # A utility function to find the vertex with  
    # minimum distance value, from the set of vertices  
    # not yet included in shortest path tree 
    def minDistance(self, dist, sptSet): 
  
        # Initilaize minimum distance for next node 
        min = sys.maxsize 
  
        # Search not nearest vertex not in the  
        # shortest path tree 
        for v in range(self.V): 
            if dist[v] < min and sptSet[v] == False: 
                min = dist[v] 
                min_index = v 
  
        return min_index 
  
    # Funtion that implements Dijkstra's single source  
    # shortest path algorithm for a graph represented  
    # using adjacency matrix representation 
    def dijkstra(self, src): 
  
        dist = [sys.maxsize] * self.V 
        dist[src] = 0
        sptSet = [False] * self.V 
  
        for cout in range(self.V): 
  
            # Pick the minimum distance vertex from  
            # the set of vertices not yet processed.  
            # u is always equal to src in first iteration 
            u = self.minDistance(dist, sptSet) 
  
            # Put the minimum distance vertex in the  
            # shotest path tree 
            sptSet[u] = True
  
            # Update dist value of the adjacent vertices  
            # of the picked vertex only if the current  
            # distance is greater than new distance and 
            # the vertex in not in the shotest path tree 
            for v in range(self.V): 
                if self.graph[u][v] > 0 and sptSet[v] == False and dist[v] > dist[u] + self.graph[u][v]: 
                        dist[v] = dist[u] + self.graph[u][v] 
  
        return dist[0] 


def main():
	# M = Número de salas
	# E = Número de corredores
	# N = Número de tubulações
	# C = Número de consultas
	M, E, N, C = (int(x) for x in input().split())

	# Lista de tuplas para armazenar os pares de vértices ligados e seu peso
	corredores_tripulante = []
	triplas = input().split()
	x = 0
	while x < 3*E:
		corredores_tripulante.append(tuple((int(triplas[x]), int(triplas[x+1]), float(triplas[x+2]))))
		x += 3

	grafo_tripulante_matriz = np.zeros((M, M))

	for corredor in corredores_tripulante:
		grafo_tripulante_matriz[corredor[0]][corredor[1]] = corredor[2]
		grafo_tripulante_matriz[corredor[1]][corredor[0]] = corredor[2]

	grafo_tripulante = Graph(M)
	grafo_tripulante.graph = grafo_tripulante_matriz

	# Lista de tuplas para o impostor será a mesma do que o tripulante adicionado as tubulações
	corredores_impostor = corredores_tripulante.copy()
	# Lista de tuplas com os vértices da tubulação
	tubulacao = []

	pares = input().split()
	x = 0
	while x < 2*N:
		tubulacao.append(tuple((int(pares[x]), int(pares[x+1]), float(1.0))))
		x += 2

	for s in range(len(tubulacao)):
		for x in range(len(corredores_impostor)):
			if corredores_impostor[x][0] == tubulacao[s][0] and corredores_impostor[x][1] == tubulacao[s][1] and corredores_impostor[x][2]>1.0:
				del corredores_impostor[x]
				corredores_impostor.insert(x, tubulacao[s])
			if corredores_impostor[x][0] == tubulacao[s][0] and corredores_impostor[x][1] == tubulacao[s][1] and corredores_impostor[x][2]<1.0:
				del tubulacao[s]
				tubulacao.insert(s, corredores_impostor[x])

	for tubo in tubulacao:
		if not tubo in corredores_impostor:
			corredores_impostor.append(tubo)

	grafo_impostor_matriz = np.zeros((M, M))

	for corredor in corredores_impostor:
		grafo_impostor_matriz[corredor[0]][corredor[1]] = corredor[2]
		grafo_impostor_matriz[corredor[1]][corredor[0]] = corredor[2]		

	grafo_impostor = Graph(M)
	grafo_impostor.graph = grafo_impostor_matriz

	# Lista de casos 
	casos = []
	x = 0
	while x < C:
		try:
			caso = input()
			casos.append(int(caso))
		except EOFError:
			None
		x += 1

	for caso in casos:
		custo_tripulante = grafo_tripulante.dijkstra(caso)
		#print(f'custo do tripulante: {custo_tripulante}')

		custo_impostor = grafo_impostor.dijkstra(caso)
		#print(f'custo do impostor  : {custo_impostor}')


		if custo_tripulante <= custo_impostor:
			print('victory')
		else:
			print('defeat')
	
if __name__ == "__main__":
	start_time = time.time()
	main()
	print("--- %s seconds ---" % (time.time() - start_time))

File: amongos/amongos.cpp
/*
Trabalho Em Trenós - PAA ENPE 2020

Thiago de Moraes Teixeira
RA: 760667
*/

#include <cstdio>
#include <iostream>
#include <limits.h> 
#include <stdio.h> 
using namespace std; 
 
// Função que acha o vértice com a menor distância dos vértices que não foram
// incluidos no caminho mínimo ainda
int minDistance(float dist[], int M, bool sptSet[]){  
    int min = 10000, min_index; 
  
    for (int v = 0; v < M; v++) 
        if (sptSet[v] == false && dist[v] <= min) 
            min = dist[v], min_index = v; 
  
    return min_index; 
} 

// Função que implementa o algoritmo de Dijkstra para um grafo representado
// pela matriz de adjacência
float dijkstra(float **graph, int M, int src){ 
	// Vetor de saída com o valor do caminho mínimo para todos os vértices partindo da fonte
    float *dist = new float[M];
  
  	// sptSet[i] será true se o vértice i estiver no caminho mínimo entre a fonte e i 
    bool *sptSet = new bool[M]; 
  
    // Inicializa todas as distâncias como ""infinito"" and stpSet[] as false 
    for (int i = 0; i < M; i++) 
        dist[i] = 10000000000, sptSet[i] = false; 
  
    // A distância do vértice pra ele mesmo sempre é 0  
    dist[src] = 0.0; 
  
    // Acha o caminho mínimo para todos os vértices a partir da fonte
    for (int count = 0; count < M - 1; count++) { 
        // Pega o vértice de distância mínima dos vértices que ainda
        // não foram processados
        int u = minDistance(dist, M, sptSet); 
  
        // Marca o vértice como processado
        sptSet[u] = true; 
  
        // Atualiza o valor da distância dos vértices adjacentes ao escolhido
        for (int v = 0; v < M; v++) 
            // Atualiza dist[v]: 
            //	- se não estiver em sptSet
            //	- se houver outro caminho de u até v e o total do caminho da 
            //		fonte para v passando por u é menor que o valor de dist[v]
            if (!sptSet[v] && graph[u][v] && dist[u] != 10000 && dist[u] + graph[u][v] < dist[v]) 
                dist[v] = dist[u] + graph[u][v]; 
    } 
  
    // Retorna o valor do menor caminho para o vértice 0 (sala do botão do nosso problema) 
    return dist[0];
}  
  
int main(){
	// M = Número de salas
	// E = Número de corredores
	// N = Número de tubulações
	// C = Número de consultas
	int M, E, N, C;
	scanf("%d %d %d %d", &M, &E, &N, &C);

	// Matriz MxM para armazenar o peso das arestas do grafo do tripulante
	float **corredores_tripulante;
	corredores_tripulante = new float *[M];
	for(int i=0; i<M; i++)
		corredores_tripulante[i] = new float[M];

	// Recebe a linha/coluna que serão os vértices e o peso da aresta entre eles
	int x = 0;
	int linha, coluna;
	float peso;
	while (x<E){
		scanf("%d %d %f", &linha, &coluna, &peso);
		corredores_tripulante[linha][coluna] = peso;
		corredores_tripulante[coluna][linha] = peso;
		x += 1;
	}

	// Matriz MxM para armazenar o peso dos vértices do grafo do impostor
	float **corredores_impostor;
	corredores_impostor = new float *[M];
	for(int i=0; i<M; i++)
		corredores_impostor[i] = new float[M];

	// Copia os valores do grafo do tripulante para o impostor
	// pois o grafo do impostor será menor(ou igual) ao grafo do tripulante
	for(int i=0; i<M; i++)
		for(int j=0; j<M; j++)
			corredores_impostor[i][j] = corredores_tripulante[i][j];

	// Matriz MxM para representar os vértices que tem tubulação entre eles
	float **tubulacao;
	tubulacao = new float *[M];
	for(int i=0; i<M; i++)
		tubulacao[i] = new float[M];	

	// Recebe a linha/coluna que são os vértices da tubulação, peso constante = 1.0 
	x = 0;
	while (x<N){
		scanf("%d %d", &linha, &coluna);
		tubulacao[linha][coluna] = 1.0;
		tubulacao[coluna][linha] = 1.0;
		x += 1;
	}

	// Varre a matriz do impostor com o propósito de:
	// - Substituir onde o valor do corredor é maior que 1.0 e há uma tubulação possível
	// - "Criar" a aresta onde não há corredor porém o impostor pode usar a tubulação, se existir
	for(int i=0; i<M; i++)
		for(int j=0; j<M; j++){
			if(tubulacao[i][j] == 1.0 && (corredores_impostor[i][j] > tubulacao[i][j] || corredores_impostor[i][j] == 0))
				corredores_impostor[i][j] = 1.0;
		}

	// Vetor com os casos de teste
	int *casos = new int[C];
	x = 0;
	while (x<C){
		scanf("%d", &casos[x]);
		x+=1;
	}

	// Varre o vetor de casos efetuando os testes e imprime o resultado
	float custo_tripulante, custo_impostor;
	for(int i=0; i<C; i++){
		custo_tripulante = dijkstra(corredores_tripulante, M, casos[i]) * 100;
		custo_impostor = dijkstra(corredores_impostor, M, casos[i]) * 100 + 0.1;

		if (custo_tripulante <= custo_impostor)
			printf("victory\n");
		else
			printf("defeat\n");
	}

    return 0; 
} 
